### PR TITLE
Only merge model weights in LoRA recipe when `save_adapter_weights_only=False`

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog

When saving LoRA checkpoints, we only need to copy and merge the full model state dict when `save_adapter_weights_only=False`. For distributed recipes, this change will only skip the merging logic.

I've also added tests for resuming recipe state with LoRA DPO.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

File: tests/recipes/test_lora_dpo_single_device.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (ebsmothers) [inline tests/recipes/test_lora_dpo_single_device.py:1]: You're a hero

Comment (SalmanMohammadi) [reply tests/recipes/test_lora_dpo_single_device.py:1]: I was just afraid someone would ask me to run the recipe manually, so this felt like less effort.

File: recipes/lora_finetune_single_device.py [hunk 2]
```diff
@@ -527,24 +529,34 @@ def save_checkpoint(self, epoch: int) -> None:
...
-        )
-        ckpt_dict.update({training.MODEL_KEY: merged_state_dict})
+            ckpt_dict.update({training.MODEL_KEY: merged_state_dict})
+        else:
+            # No need to merge state dict if we're only saving adapter weights
+            adapter_state_dict = {
+                k: v
+                for k, v in self._model.state_dict().items()
+                if adapter_key_filter(k)
+            }
```
Comment (ebsmothers) [inline recipes/lora_finetune_single_device.py:559]: Doesn't this still materialize the entire state dict though?

File: recipes/lora_finetune_distributed.py [hunk 3]
```diff
@@ -628,12 +628,13 @@ def save_checkpoint(
             checkpoint_dict.update({training.ADAPTER_KEY: adapter_state_dict})
 
             # merge the adapter weights and base weights to create the model checkpoint
-            merged_state_dict = get_merged_lora_ckpt(
-                cpu_state_dict,
-                rank=self._lora_rank,
-                alpha=self._lora_alpha,
-            )
-            checkpoint_dict.update({training.MODEL_KEY: merged_state_dict})
+            if not self._save_adapter_weights_only:
```
Comment (ebsmothers) [inline recipes/lora_finetune_distributed.py:631]: I think @pbontrager had looked at something with this previously, but we still create the full state dict on CPU in this case then, right? To actually reduce the memory footprint to just the adapter weights we'd need to change the call to `get_full_model_state_dict` above (btw it's ok to say "let's do this in a follow-up")

**Review by ebsmothers — APPROVED** [review]
A couple questions around whether we can get around materializing the full model state dict when `save_adapter_weights_only=True`. (I think it'll be trickier for the distributed recipe, but would be nice to sort out for single-device.) Other than that looks good!
